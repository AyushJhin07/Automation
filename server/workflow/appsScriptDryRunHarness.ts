import assert from 'node:assert';
import { readFile, readdir } from 'node:fs/promises';
import { basename, extname, resolve } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import crypto from 'node:crypto';
import util from 'node:util';
import vm from 'node:vm';

import type { WorkflowGraph } from '../../common/workflow-types';
import { compileToAppsScript } from './compile-to-appsscript';

export interface AppsScriptFixture {
  id: string;
  description?: string;
  graph: WorkflowGraph;
  entry?: {
    context?: Record<string, any>;
  };
  secrets?: Record<string, string>;
  http?: HttpFixture[];
  expect?: FixtureExpectations;
}

export interface FixtureExpectations {
  context?: Record<string, any>;
  logs?: LogExpectation[];
  httpCalls?: HttpCallExpectation[];
}

export interface HttpFixture {
  name?: string;
  request: HttpRequestExpectation;
  response: HttpResponseMock;
}

export interface HttpRequestExpectation {
  url: string;
  method?: string;
  headers?: Record<string, string>;
  payload?: any;
}

export interface HttpResponseMock {
  status: number;
  body?: any;
  headers?: Record<string, string>;
}

export interface LogExpectation {
  level?: 'log' | 'warn' | 'error';
  includes?: string;
  matches?: string;
}

export interface HttpCallExpectation {
  url?: string;
  method?: string;
  includesPayloadFragment?: string;
}

export interface LogEntry {
  level: 'log' | 'warn' | 'error';
  message: string;
  args: unknown[];
}

export interface RecordedHttpCall {
  url: string;
  method: string;
  headers: Record<string, string>;
  payload?: any;
  fixtureName?: string;
}

export interface FixtureRunResult {
  id: string;
  description?: string;
  success: boolean;
  durationMs: number;
  context?: any;
  logs: LogEntry[];
  httpCalls: RecordedHttpCall[];
  error?: string;
  failedExpectations?: string[];
  stack?: string;
}

export interface AppsScriptDryRunSummary {
  results: FixtureRunResult[];
  passed: number;
  failed: number;
  durationMs: number;
}

interface SandboxOptions {
  secrets?: Record<string, string>;
  httpFixtures?: HttpFixture[];
}

interface SandboxRunResult {
  context: any;
  logs: LogEntry[];
  httpCalls: RecordedHttpCall[];
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = resolve(__filename, '..');
const DEFAULT_FIXTURE_DIR = resolve(__dirname, '__tests__', 'apps-script-fixtures');

class FixtureAssertionError extends Error {
  public readonly expectationFailures: string[];
  public readonly details?: { context?: any; logs?: LogEntry[]; httpCalls?: RecordedHttpCall[] };

  constructor(message: string, failures: string[], details?: { context?: any; logs?: LogEntry[]; httpCalls?: RecordedHttpCall[] }) {
    super(message);
    this.expectationFailures = failures;
    this.details = details;
    this.name = 'FixtureAssertionError';
  }
}

class ConsoleCapture {
  public readonly logs: LogEntry[] = [];

  log = (...args: unknown[]) => {
    this.push('log', args);
  };

  warn = (...args: unknown[]) => {
    this.push('warn', args);
  };

  error = (...args: unknown[]) => {
    this.push('error', args);
  };

  private push(level: 'log' | 'warn' | 'error', args: unknown[]): void {
    const message = args.length === 0 ? '' : util.format(...args);
    this.logs.push({ level, message, args });
  }
}

class HttpResponse {
  constructor(private readonly mock: HttpResponseMock) {}

  getContentText(): string {
    if (typeof this.mock.body === 'string') {
      return this.mock.body;
    }
    if (this.mock.body === undefined || this.mock.body === null) {
      return '';
    }
    return JSON.stringify(this.mock.body);
  }

  getResponseCode(): number {
    return this.mock.status;
  }

  getAllHeaders(): Record<string, string> {
    return { ...(this.mock.headers ?? {}) };
  }
}

class UrlFetchStub {
  private readonly queue: HttpFixture[];
  private readonly callsInternal: RecordedHttpCall[] = [];

  constructor(fixtures: HttpFixture[]) {
    this.queue = fixtures.map(fixture => ({ ...fixture }));
  }

  get api() {
    return {
      fetch: (url: string, options?: any) => this.fetch(url, options)
    };
  }

  get calls(): RecordedHttpCall[] {
    return this.callsInternal;
  }

  private normalizeHeaders(headers?: Record<string, string> | null): Record<string, string> {
    const normalized: Record<string, string> = {};
    if (!headers) {
      return normalized;
    }
    for (const [key, value] of Object.entries(headers)) {
      normalized[key.toLowerCase()] = value;
    }
    return normalized;
  }

  private normalizePayload(payload: any): any {
    if (payload === undefined || payload === null) {
      return undefined;
    }
    if (typeof payload === 'string') {
      try {
        return JSON.parse(payload);
      } catch {
        return payload;
      }
    }
    if (payload instanceof ArrayBuffer || ArrayBuffer.isView(payload)) {
      return this.normalizePayload(Buffer.from(payload as any).toString('utf8'));
    }
    if (typeof payload === 'object') {
      return payload;
    }
    return payload;
  }

  fetch(url: string, rawOptions?: any): HttpResponse {
    if (this.queue.length === 0) {
      throw new Error(`Unexpected UrlFetchApp.fetch call for ${url}`);
    }

    const fixture = this.queue.shift()!;
    const options = rawOptions && typeof rawOptions === 'object' ? rawOptions : {};
    const method = (options.method || 'GET').toString().toUpperCase();
    const headers = this.normalizeHeaders(options.headers ?? (options.header || {}));
    const payload = this.normalizePayload(options.payload ?? options.body);

    if (fixture.request.url !== url) {
      throw new Error(`Expected HTTP request to ${fixture.request.url} but received ${url}`);
    }

    if (fixture.request.method && fixture.request.method.toUpperCase() !== method) {
      throw new Error(`Expected HTTP method ${fixture.request.method.toUpperCase()} but received ${method}`);
    }

    if (fixture.request.headers) {
      for (const [key, expected] of Object.entries(fixture.request.headers)) {
        const actual = headers[key.toLowerCase()];
        assert.strictEqual(
          actual,
          expected,
          `Expected header ${key}=${expected} but received ${actual ?? 'undefined'} on ${fixture.request.url}`
        );
      }
    }

    if (fixture.request.payload !== undefined) {
      assertSubset(
        payload,
        fixture.request.payload,
        `payload for ${fixture.request.url}`
      );
    }

    this.callsInternal.push({
      url,
      method,
      headers,
      payload,
      fixtureName: fixture.name
    });

    return new HttpResponse(fixture.response);
  }

  assertAllConsumed(): void {
    if (this.queue.length > 0) {
      const remaining = this.queue.map(f => f.request.url).join(', ');
      throw new Error(`Expected HTTP interactions not executed: ${remaining}`);
    }
  }
}

class PropertiesStore {
  private readonly store: Record<string, string>;

  constructor(initial: Record<string, string>) {
    this.store = { ...initial };
  }

  getProperty(key: string): string | null {
    return Object.prototype.hasOwnProperty.call(this.store, key) ? this.store[key] : null;
  }

  setProperty(key: string, value: any): void {
    this.store[key] = value == null ? '' : String(value);
  }

  deleteProperty(key: string): void {
    delete this.store[key];
  }

  getProperties(): Record<string, string> {
    return { ...this.store };
  }
}

export class AppsScriptSandbox {
  private readonly consoleCapture = new ConsoleCapture();
  private readonly urlFetch: UrlFetchStub;
  private readonly properties: PropertiesStore;
  private readonly context: vm.Context;

  constructor(options: SandboxOptions) {
    this.urlFetch = new UrlFetchStub(options.httpFixtures ?? []);
    this.properties = new PropertiesStore(options.secrets ?? {});

    const utilities = this.createUtilities();

    const scriptProperties = {
      getProperty: (key: string) => this.properties.getProperty(key),
      setProperty: (key: string, value: any) => {
        this.properties.setProperty(key, value);
      },
      deleteProperty: (key: string) => {
        this.properties.deleteProperty(key);
      },
      getProperties: () => this.properties.getProperties(),
    };

    const propertiesService = {
      getScriptProperties: () => scriptProperties,
      getUserProperties: () => scriptProperties,
      getDocumentProperties: () => scriptProperties,
    };

    const logger = {
      log: (...args: unknown[]) => this.consoleCapture.log(...args),
      info: (...args: unknown[]) => this.consoleCapture.log(...args),
      warn: (...args: unknown[]) => this.consoleCapture.warn(...args),
      error: (...args: unknown[]) => this.consoleCapture.error(...args),
    };

    const session = {
      getActiveUser: () => ({
        getEmail: () => 'apps-script-emulator@example.com'
      })
    };

    const sandbox: Record<string, any> = {
      console: this.consoleCapture,
      Logger: logger,
      PropertiesService: propertiesService,
      UrlFetchApp: this.urlFetch.api,
      Utilities: utilities,
      Session: session,
      ScriptApp: this.createScriptAppStub(),
      globalThis: undefined,
      global: undefined,
    };

    sandbox.globalThis = sandbox;
    sandbox.global = sandbox;

    this.context = vm.createContext(sandbox, {
      name: 'AppsScriptSandbox'
    });
  }

  evaluate(code: string): void {
    const script = new vm.Script(code, { filename: 'Code.gs' });
    script.runInContext(this.context, { displayErrors: true });
  }

  async runMain(initialContext: Record<string, any>): Promise<SandboxRunResult> {
    const mainFn = (this.context as any).main;
    if (typeof mainFn !== 'function') {
      throw new Error('Compiled Apps Script bundle did not expose a main(ctx) function');
    }

    const cloned = clone(initialContext ?? {});
    const maybePromise = mainFn(cloned);
    const contextResult = await Promise.resolve(maybePromise);

    return {
      context: contextResult,
      logs: this.consoleCapture.logs,
      httpCalls: this.urlFetch.calls,
    };
  }

  async runFunction(functionName: string, ...args: any[]): Promise<SandboxRunResult> {
    if (!functionName || typeof functionName !== 'string') {
      throw new Error('runFunction requires a function name');
    }

    const target = (this.context as any)[functionName];
    if (typeof target !== 'function') {
      throw new Error(`Compiled Apps Script bundle did not expose a ${functionName} function`);
    }

    const result = await Promise.resolve(target.apply(undefined, args));

    return {
      context: result,
      logs: this.consoleCapture.logs,
      httpCalls: this.urlFetch.calls,
    };
  }

  verifyHttpExpectations(): void {
    this.urlFetch.assertAllConsumed();
  }

  private createUtilities(): Record<string, any> {
    return {
      sleep: (_ms: number) => { /* no-op in emulator */ },
      base64Decode: (value: string) => Buffer.from(value, 'base64'),
      base64Encode: (value: any) => Buffer.from(value).toString('base64'),
      base64EncodeWebSafe: (value: any) => Buffer.from(value).toString('base64url'),
      computeHmacSha256: (data: any, key: any) => {
        const dataBuffer = Array.isArray(data) ? Buffer.from(data) : Buffer.from(data);
        const keyBuffer = Array.isArray(key) ? Buffer.from(key) : Buffer.from(key);
        return crypto.createHmac('sha256', keyBuffer).update(dataBuffer).digest();
      },
      newBlob: (value: any, _contentType?: string) => {
        const buffer = Buffer.isBuffer(value) ? value : Buffer.from(String(value ?? ''));
        return {
          getBytes: () => Array.from(buffer),
        };
      },
      formatDate: (date: Date | number | string, _timezone: string, _format: string) => {
        const d = typeof date === 'string' || typeof date === 'number' ? new Date(date) : date;
        return d.toISOString();
      },
      getUuid: () => crypto.randomUUID(),
    };
  }

  private createScriptAppStub(): Record<string, any> {
    const builder = {
      timeBased: () => ({
        everyMinutes: (_minutes: number) => ({ create: () => ({ getUniqueId: () => 'trigger-' + Date.now() }) }),
        everyHours: (_hours: number) => ({ create: () => ({ getUniqueId: () => 'trigger-' + Date.now() }) }),
        atHour: (_hour: number) => ({ onWeekDay: () => ({ create: () => ({ getUniqueId: () => 'trigger-' + Date.now() }) }) }),
        create: () => ({ getUniqueId: () => 'trigger-' + Date.now() })
      }),
      forSpreadsheet: () => ({ onEdit: () => ({ create: () => ({ getUniqueId: () => 'trigger-' + Date.now() }) }) })
    };

    return {
      getProjectTriggers: () => [],
      deleteTrigger: () => {},
      newTrigger: () => builder,
      WeekDay: {
        MONDAY: 'MONDAY',
        TUESDAY: 'TUESDAY',
        WEDNESDAY: 'WEDNESDAY',
        THURSDAY: 'THURSDAY',
        FRIDAY: 'FRIDAY',
        SATURDAY: 'SATURDAY',
        SUNDAY: 'SUNDAY'
      },
    };
  }
}

function clone<T>(value: T): T {
  if (typeof structuredClone === 'function') {
    return structuredClone(value);
  }
  return JSON.parse(JSON.stringify(value));
}

function assertSubset(actual: any, expected: any, path: string = ''): void {
  if (expected === null || typeof expected !== 'object') {
    assert.deepStrictEqual(actual, expected, `Mismatch at ${path || 'value'}`);
    return;
  }

  if (Array.isArray(expected)) {
    assert(Array.isArray(actual), `Expected array at ${path || 'value'}`);
    assert.strictEqual(actual.length, expected.length, `Array length mismatch at ${path || 'value'}`);
    expected.forEach((item, index) => {
      assertSubset(actual[index], item, `${path}[${index}]`);
    });
    return;
  }

  assert(actual && typeof actual === 'object', `Expected object at ${path || 'value'}`);

  for (const [key, value] of Object.entries(expected)) {
    const childPath = path ? `${path}.${key}` : key;
    assertSubset((actual as any)[key], value, childPath);
  }
}

function ensureLogExpectations(logs: LogEntry[], expectations: LogExpectation[]): string[] {
  const failures: string[] = [];
  for (const expectation of expectations) {
    const matcher = (entry: LogEntry) => {
      if (expectation.level && entry.level !== expectation.level) {
        return false;
      }
      if (expectation.includes && !entry.message.includes(expectation.includes)) {
        return false;
      }
      if (expectation.matches) {
        const regex = new RegExp(expectation.matches);
        if (!regex.test(entry.message)) {
          return false;
        }
      }
      return true;
    };

    if (!logs.some(matcher)) {
      failures.push(
        expectation.includes
          ? `Expected log containing "${expectation.includes}" not found`
          : expectation.matches
            ? `Expected log matching /${expectation.matches}/ not found`
            : 'Expected log entry not found'
      );
    }
  }
  return failures;
}

function ensureHttpCallExpectations(calls: RecordedHttpCall[], expectations: HttpCallExpectation[]): string[] {
  const failures: string[] = [];
  for (const expectation of expectations) {
    const matcher = (call: RecordedHttpCall) => {
      if (expectation.url && call.url !== expectation.url) {
        return false;
      }
      if (expectation.method && call.method !== expectation.method.toUpperCase()) {
        return false;
      }
      if (expectation.includesPayloadFragment) {
        const payloadString = typeof call.payload === 'string' ? call.payload : JSON.stringify(call.payload ?? '');
        if (!payloadString.includes(expectation.includesPayloadFragment)) {
          return false;
        }
      }
      return true;
    };

    if (!calls.some(matcher)) {
      failures.push(
        `Expected HTTP call${expectation.url ? ` to ${expectation.url}` : ''} ` +
        `${expectation.method ? `with method ${expectation.method.toUpperCase()} ` : ''}`.trim()
      );
    }
  }
  return failures;
}

async function executeFixture(fixture: AppsScriptFixture): Promise<FixtureRunResult> {
  const start = Date.now();
  const compiled = compileToAppsScript(fixture.graph);
  const codeFile = compiled.files.find(file => file.path === 'Code.gs');

  if (!codeFile) {
    throw new Error(`Fixture ${fixture.id} did not produce a Code.gs file`);
  }

  const sandbox = new AppsScriptSandbox({
    secrets: fixture.secrets,
    httpFixtures: fixture.http ?? [],
  });

  sandbox.evaluate(codeFile.content);

  const { context, logs, httpCalls } = await sandbox.runMain(fixture.entry?.context ?? {});

  const expectationFailures: string[] = [];

  try {
    sandbox.verifyHttpExpectations();
  } catch (error) {
    expectationFailures.push(error instanceof Error ? error.message : String(error));
  }

  if (fixture.expect?.context) {
    try {
      assertSubset(context, fixture.expect.context, 'context');
    } catch (error: any) {
      expectationFailures.push(error?.message ?? String(error));
    }
  }

  if (fixture.expect?.logs) {
    expectationFailures.push(...ensureLogExpectations(logs, fixture.expect.logs));
  }

  if (fixture.expect?.httpCalls) {
    expectationFailures.push(...ensureHttpCallExpectations(httpCalls, fixture.expect.httpCalls));
  }

  if (expectationFailures.length > 0) {
    throw new FixtureAssertionError(
      `Fixture ${fixture.id} failed expectations`,
      expectationFailures,
      { context, logs, httpCalls }
    );
  }

  return {
    id: fixture.id,
    description: fixture.description,
    success: true,
    durationMs: Date.now() - start,
    context,
    logs,
    httpCalls,
  };
}

export async function loadAppsScriptFixtures(dir: string = DEFAULT_FIXTURE_DIR): Promise<AppsScriptFixture[]> {
  const entries = await readdir(dir);
  const fixtures: AppsScriptFixture[] = [];

  for (const entry of entries) {
    if (extname(entry) !== '.json') {
      continue;
    }

    const filePath = resolve(dir, entry);
    const raw = await readFile(filePath, 'utf-8');
    const parsed = JSON.parse(raw) as AppsScriptFixture;
    if (!parsed.id) {
      parsed.id = basename(entry, '.json');
    }
    fixtures.push(parsed);
  }

  return fixtures.sort((a, b) => a.id.localeCompare(b.id));
}

export interface RunFixturesOptions {
  fixturesDir?: string;
  filterIds?: string[];
  stopOnError?: boolean;
}

export async function runAppsScriptFixtures(options: RunFixturesOptions = {}): Promise<AppsScriptDryRunSummary> {
  const fixtures = await loadAppsScriptFixtures(options.fixturesDir ?? DEFAULT_FIXTURE_DIR);
  const filterSet = options.filterIds && options.filterIds.length > 0
    ? new Set(options.filterIds)
    : null;

  const results: FixtureRunResult[] = [];
  const start = Date.now();

  for (const fixture of fixtures) {
    if (filterSet && !filterSet.has(fixture.id)) {
      continue;
    }

    const fixtureStart = Date.now();

    try {
      const result = await executeFixture(fixture);
      results.push(result);
    } catch (error: any) {
      const errorDetails = error instanceof FixtureAssertionError ? error.details : undefined;
      const failure: FixtureRunResult = {
        id: fixture.id,
        description: fixture.description,
        success: false,
        durationMs: Date.now() - fixtureStart,
        logs: errorDetails?.logs ?? [],
        httpCalls: errorDetails?.httpCalls ?? [],
        error: error?.message ?? String(error),
        failedExpectations: error instanceof FixtureAssertionError ? error.expectationFailures : undefined,
        stack: error?.stack,
      };
      if (errorDetails?.context) {
        failure.context = errorDetails.context;
      }
      results.push(failure);
      if (options.stopOnError) {
        break;
      }
    }
  }

  const passed = results.filter(result => result.success).length;
  const failed = results.length - passed;

  return {
    results,
    passed,
    failed,
    durationMs: Date.now() - start,
  };
}

export async function runSingleFixture(fixtureId: string, fixturesDir?: string): Promise<FixtureRunResult> {
  const summary = await runAppsScriptFixtures({ fixturesDir, filterIds: [fixtureId] });
  if (summary.results.length === 0) {
    throw new Error(`No fixture found for id ${fixtureId}`);
  }
  return summary.results[0];
}

export async function importFixtureModule(modulePath: string): Promise<AppsScriptFixture> {
  const moduleUrl = pathToFileURL(resolve(modulePath)).href;
  const mod = await import(moduleUrl);
  if (!mod.default) {
    throw new Error(`Fixture module ${modulePath} does not export default fixture definition`);
  }
  return mod.default as AppsScriptFixture;
}
