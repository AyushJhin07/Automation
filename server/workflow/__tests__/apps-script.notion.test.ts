import { describe, expect, it } from 'vitest';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { REAL_OPS } from '../compile-to-appsscript';
import { runSingleFixture } from '../appsScriptDryRunHarness';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const fixturesDir = path.join(__dirname, 'apps-script-fixtures');

describe('Apps Script Notion REAL_OPS', () => {
  it('builds action.notion:create_page', () => {
    const config = {
      parent: { type: 'database_id' },
      properties: {
        Name: {
          title: [
            {
              text: { content: '{{pageTitle}}' }
            }
          ]
        },
        Status: {
          select: { name: 'In Progress' }
        }
      },
      children: [
        {
          type: 'paragraph',
          paragraph: {
            text: [
              {
                type: 'text',
                text: { content: 'Automation generated content.' }
              }
            ]
          }
        }
      ],
      icon: { type: 'emoji', emoji: 'âš¡' },
      cover: { type: 'external', external: { url: 'https://example.com/cover.png' } }
    };

    expect(REAL_OPS['action.notion:create_page'](config)).toMatchSnapshot();
  });
});

describe('Apps Script Notion integration', () => {
  it('creates a page via the Notion REST API', async () => {
    const result = await runSingleFixture('notion-create-page', fixturesDir);

    expect(result.success).toBe(true);
    expect(result.context.notionPageId).toBe('page-123');
    expect(result.context.notionPageUrl).toBe('https://www.notion.so/Test-Page');
    expect(result.context.notionCreatePageResponse).toMatchObject({
      status: 200,
      requestId: 'req-notion-1'
    });
    expect(result.httpCalls).toHaveLength(1);
    expect(result.httpCalls[0].url).toBe('https://api.notion.com/v1/pages');
  });
});
