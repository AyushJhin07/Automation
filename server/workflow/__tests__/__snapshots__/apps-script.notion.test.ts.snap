exports[`Apps Script Notion REAL_OPS builds action.notion:create_page 1`] = `
function step_action_notion_create_page(ctx) {
  ctx = ctx || {};

  const accessToken = requireOAuthToken('notion', { scopes: ['read_content', 'update_content', 'insert_content'] });

  const parentConfig = {"type":"database_id"};
  const propertiesConfig = {"Name":{"title":[{"text":{"content":"{{pageTitle}}"}}]},"Status":{"select":{"name":"In Progress"}}};
  const childrenConfig = [{"type":"paragraph","paragraph":{"text":[{"type":"text","text":{"content":"Automation generated content."}}]}}];
  const iconConfig = {"type":"emoji","emoji":"âš¡"};
  const coverConfig = {"type":"external","external":{"url":"https://example.com/cover.png"}};

  function optionalSecret(name) {
    try {
      const value = getSecret(name, { connectorKey: 'notion' });
      return typeof value === 'string' ? value.trim() : String(value || '').trim();
    } catch (error) {
      return '';
    }
  }

  function resolveString(template, options) {
    if (template === null || template === undefined) {
      return '';
    }
    if (typeof template !== 'string') {
      return String(template);
    }
    const trimmed = template.trim();
    if (!trimmed && options && options.allowEmpty) {
      return '';
    }
    const resolved = interpolate(trimmed, ctx).trim();
    if (!resolved && options && options.fallbackSecret) {
      const fallback = optionalSecret(options.fallbackSecret);
      if (fallback) {
        return fallback;
      }
    }
    return resolved;
  }

  function resolveStructured(value) {
    if (value === null || value === undefined) {
      return undefined;
    }
    if (Array.isArray(value)) {
      const result = [];
      for (let i = 0; i < value.length; i++) {
        result.push(resolveStructured(value[i]));
      }
      return result;
    }
    if (typeof value === 'object') {
      const result = {};
      for (const key in value) {
        if (Object.prototype.hasOwnProperty.call(value, key)) {
          result[key] = resolveStructured(value[key]);
        }
      }
      return result;
    }
    if (typeof value === 'string') {
      return interpolate(value, ctx);
    }
    return value;
  }

  function ensureParent(config) {
    if (!config || typeof config !== 'object') {
      throw new Error('Notion create_page requires a parent configuration.');
    }
    const type = typeof config.type === 'string' ? config.type.trim().toLowerCase() : '';
    if (!type) {
      throw new Error("Notion create_page requires parent.type set to 'database_id', 'page_id', or 'workspace'.");
    }

    if (type === 'database_id') {
      const raw = config.database_id !== undefined ? config.database_id : (config.databaseId !== undefined ? config.databaseId : null);
      const databaseId = resolveString(raw ?? '', { fallbackSecret: 'NOTION_DATABASE_ID' });
      if (!databaseId) {
        throw new Error('Notion create_page requires a database_id in the manifest or the NOTION_DATABASE_ID script property.');
      }
      return { database_id: databaseId };
    }

    if (type === 'page_id') {
      const raw = config.page_id !== undefined ? config.page_id : (config.pageId !== undefined ? config.pageId : null);
      const pageId = resolveString(raw ?? '', { fallbackSecret: 'NOTION_PAGE_ID' });
      if (!pageId) {
        throw new Error('Notion create_page requires a page_id in the manifest or the NOTION_PAGE_ID script property.');
      }
      return { page_id: pageId };
    }

    if (type === 'workspace') {
      return { workspace: true };
    }

    throw new Error('Unsupported Notion parent type: ' + type + '.');
  }

  const parent = ensureParent(parentConfig);
  const properties = resolveStructured(propertiesConfig);
  if (!properties || Object.keys(properties).length === 0) {
    throw new Error('Notion create_page requires at least one property. Configure the Properties block in the manifest.');
  }

  const requestBody = {
    parent: parent,
    properties: properties
  };

  const children = resolveStructured(childrenConfig);
  if (children && Array.isArray(children) && children.length > 0) {
    requestBody.children = children;
  }

  const icon = resolveStructured(iconConfig);
  if (icon && typeof icon === 'object' && Object.keys(icon).length > 0) {
    requestBody.icon = icon;
  }

  const cover = resolveStructured(coverConfig);
  if (cover && typeof cover === 'object' && Object.keys(cover).length > 0) {
    requestBody.cover = cover;
  }

  try {
    const response = rateLimitAware(
      () => fetchJson({
        url: 'https://api.notion.com/v1/pages',
        method: 'POST',
        headers: {
          'Authorization': 'Bearer ' + accessToken,
          'Content-Type': 'application/json',
          'Notion-Version': '2022-06-28'
        },
        payload: JSON.stringify(requestBody),
        contentType: 'application/json'
      }),
      { attempts: 4, initialDelayMs: 750, maxDelayMs: 6000, jitter: 0.25 }
    );

    const page = response.body || {};
    const headers = response.headers || {};
    const requestId = headers['x-request-id'] || headers['X-Request-Id'] || null;

    ctx.notionPageId = page.id || null;
    ctx.notionPageUrl = page.url || null;
    ctx.notionPage = page;
    ctx.notionCreatePageResponse = {
      status: response.status,
      requestId: requestId,
      parent: parent,
      headers: headers,
      body: page
    };

    logInfo('notion_create_page_success', {
      pageId: ctx.notionPageId || null,
      parentType: parent.database_id ? 'database_id' : parent.page_id ? 'page_id' : 'workspace',
      requestId: requestId || undefined
    });

    return ctx;
  } catch (error) {
    const status = error && typeof error.status === 'number' ? error.status : null;
    const body = error && Object.prototype.hasOwnProperty.call(error, 'body') ? error.body : null;
    const providerCode = body && typeof body === 'object' ? body.code || null : null;
    const providerMessage = body && typeof body === 'object' && body.message ? body.message : (error && error.message ? error.message : String(error));

    logError('notion_create_page_failed', {
      status: status,
      providerCode: providerCode,
      message: providerMessage
    });

    throw new Error('Notion create_page failed: ' + (providerCode ? providerCode + ' ' : '') + providerMessage);
  }
}`;
